File: perl5/TEST_A~1.pl
#!/usr/bin/perl
# TEST_A~1.PY — portiert nach perl5
# Quelle: python, OpenClaw@gateway1:abstraction-manager/TEST_A~1.PY
# auch in: Projects@abstractions:abstractions/test_abstractions_manager.py
# Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use File::Temp qw(tempdir);
use File::Spec;
use File::Path qw(make_path);
use Digest::SHA qw(sha256_hex);
use JSON qw(decode_json encode_json);
use Test::More;
use Test::Exception;
use FindBin qw($Bin);
use lib $Bin;

# Simuliere Exceptions
package ValidationError {
    sub new {
        my ($class, $message) = @_;
        return bless { message => $message }, $class;
    }
    sub message { $_[0]->{message} }
}

package PortationError {
    sub new {
        my ($class, $message) = @_;
        return bless { message => $message }, $class;
    }
    sub message { $_[0]->{message} }
}

package ApiKeyError {
    sub new {
        my ($class, $message) = @_;
        return bless { message => $message }, $class;
    }
    sub message { $_[0]->{message} }
}

package StateFileError {
    sub new {
        my ($class, $message) = @_;
        return bless { message => $message }, $class;
    }
    sub message { $_[0]->{message} }
}

# Simuliere validators
package validators {
    our @ALLOWED_TARGET_LANGUAGES = qw(perl5 perl6 javascript python bash powershell go);
    our @ALLOWED_AI_MODELS = (
        'openrouter/anthropic/claude-3.5-sonnet:beta',
        'openrouter/google/gemini-pro-1.5',
        'openrouter/meta-llama/llama-3-70b-instruct',
        'openrouter/mistralai/mixtral-8x22b-instruct',
        'openrouter/openai/gpt-4-turbo-2024-04-09',
        'openrouter/qwen/qwen2-72b-instruct',
    );

    sub validate_source_file_path {
        my ($path) = @_;
        my $workspace = $ENV{OPENCLAW_WORKSPACE} || die "OPENCLAW_WORKSPACE nicht gesetzt";
        $path = File::Spec->rel2abs($path);
        die ValidationError->new("Datei nicht gefunden") unless -f $path;
        die ValidationError->new("kein reguläres File") if -d $path;
        die ValidationError->new("Pfad außerhalb des Workspaces") unless $path =~ /^\Q$workspace\E/;
        return $path;
    }

    sub validate_task_description {
        my ($task) = @_;
        $task =~ s/^\s+|\s+$//g;
        die ValidationError->new("Leere Task-Beschreibung") if $task eq '';
        die ValidationError->new("Task zu lang") if length($task) > 500;
        die ValidationError->new("Unerlaubte Shell-Metazeichen") if $task =~ /[;&`$()|<>]/;
        return $task;
    }

    sub validate_ai_model_name {
        my ($model) = @_;
        die ValidationError->new("Modell nicht erlaubt") unless grep { $_ eq $model } @ALLOWED_AI_MODELS;
        return $model;
    }

    sub validate_target_language {
        my ($lang) = @_;
        $lang = lc($lang);
        die ValidationError->new("Sprache nicht unterstützt") unless grep { $_ eq $lang } @ALLOWED_TARGET_LANGUAGES;
        return $lang;
    }

    sub validate_timeout_seconds {
        my ($timeout) = @_;
        die ValidationError->new("Timeout muss Integer sein") unless $timeout =~ /^\d+$/;
        die ValidationError->new("Timeout außerhalb 1–7200") unless $timeout >= 1 && $timeout <= 7200;
        return $timeout;
    }

    sub load_and_validate_api_key {
        my ($provider) = @_;
        my $key = $ENV{"${provider}_API_KEY"};
        die ApiKeyError->new("API-Key nicht gesetzt") unless defined $key;
        die ApiKeyError->new("API-Key zu kurz") if length($key) < 20;
        return $key;
    }
}

# Simuliere create_abstraction
package create_abstraction {
    our %LANGUAGE_FILE_EXTENSIONS = (
        perl5 => '.pl',
        perl6 => '.raku',
        javascript => '.js',
        python => '.py',
        bash => '.sh',
        powershell => '.ps1',
        go => '.go',
    );

    sub compute_file_sha256 {
        my ($file) = @_;
        open my $fh, '<', $file or die "Kann $file nicht öffnen: $!";
        binmode $fh;
        my $digest = sha256_hex(do { local $/; <$fh> });
        close $fh;
        return $digest;
    }

    sub has_source_file_changed {
        my ($file, $hash_cache) = @_;
        my $current_hash = compute_file_sha256($file);
        my $cached_hash = $hash_cache->{$file};
        return !defined $cached_hash || $current_hash ne $cached_hash;
    }

    sub load_abstraction_state {
        my ($state_file) = @_;
        return {} unless -f $state_file;
        open my $fh, '<', $state_file or die "Kann $state_file nicht öffnen: $!";
        my $json_text = do { local $/; <$fh> };
        close $fh;
        eval {
            return decode_json($json_text);
        } or do {
            die StateFileError->new("Kann State-Datei nicht parsen: $@");
        };
    }

    sub save_abstraction_state_atomically {
        my ($state_file, $state) = @_;
        my ($volume, $directories) = File::Spec->splitpath($state_file);
        make_path($directories) unless -d $directories;
        my $temp_file = $state_file . '.tmp';
        open my $fh, '>', $temp_file or die "Kann $temp_file nicht öffnen: $!";
        print $fh encode_json($state);
        close $fh;
        rename $temp_file, $state_file or die "Kann $temp_file nicht nach $state_file umbenennen: $!";
    }
}

# ===========================================================================
# Fixtures
# ===========================================================================

sub temp_workspace {
    my $tmp_path = tempdir(CLEANUP => 1);
    my $workspace = File::Spec->catdir($tmp_path, 'workspace');
    my $scripts_dir = File::Spec->catdir($workspace, 'skills', 'scripts');
    make_path($scripts_dir);
    return $workspace;
}

sub valid_perl_script {
    my ($temp_workspace) = @_;
    my $scripts_dir = File::Spec->catdir($temp_workspace, 'skills', 'scripts');
    my $script = File::Spec->catfile($scripts_dir, 'db_maintainer.pl');
    open my $fh, '>', $script or die "Kann $script nicht öffnen: $!";
    print $fh "# Test-Script\nprint 'hello world';\n";
    close $fh;
    return $script;
}

sub state_file_path {
    my $tmp_path = tempdir(CLEANUP => 1);
    return File::Spec->catfile($tmp_path, 'db', 'abstractions_state.json');
}

# ===========================================================================
# Tests: Path-Traversal-Schutz
# ===========================================================================

subtest 'Path-Traversal-Schutz' => sub {
    my $workspace = temp_workspace();
    my $script = valid_perl_script($workspace);
    local $ENV{OPENCLAW_WORKSPACE} = $workspace;

    lives_ok {
        validators::validate_source_file_path($script);
    } 'Gültiger Pfad innerhalb des erlaubten Verzeichnisses wird akzeptiert';

    throws_ok {
        validators::validate_source_file_path('/absolutely/nonexistent/path/script.pl');
    } 'FileNotFoundError', 'Nicht existente Datei löst FileNotFoundError aus';

    throws_ok {
        validators::validate_source_file_path($workspace);
    } 'ValidationError', 'Verzeichnis-Pfad statt Datei löst ValidationError aus';

    throws_ok {
        validators::validate_source_file_path('/etc/passwd');
    } ['ValidationError', 'FileNotFoundError'], 'Pfade außerhalb des Workspaces werden blockiert';
};

# ===========================================================================
# Tests: Shell-Injection-Prävention
# ===========================================================================

subtest 'Shell-Injection-Prävention' => sub {
    my @valid_tasks = (
        'Port db_maintainer.pl to Go',
        'Add error handling to json_processor',
        'Refactor websearch-crawl.sh for Perl 5',
        'Port check-live.js with full JSDoc',
        'Migrate log_collector.pl to Ruby',
    );

    for my $task (@valid_tasks) {
        lives_ok {
            validators::validate_task_description($task);
        } "Gültige Task-Beschreibung akzeptiert: $task";
    }

    my @malicious_inputs = (
        ['; rm -rf /',           'Semikolon + Befehl'],
        ['$(cat /etc/passwd)',   'Befehlssubstitution $(...)'],
        ['`whoami`',             'Backtick-Substitution'],
        ['task && evil_cmd',     '&&-Verkettung'],
        ['task | cat /etc/shadow', 'Pipe-Redirect'],
        ["task\ncommand",        'Newline-Injection'],
        ['task; exit 0',         'Semikolon-Injection'],
        ['task > /tmp/evil',     'Ausgabe-Redirect'],
    );

    for my $malicious (@malicious_inputs) {
        my ($input, $desc) = @$malicious;
        throws_ok {
            validators::validate_task_description($input);
        } 'ValidationError', "Shell-Metazeichen blockiert: $desc";
    }

    throws_ok {
        validators::validate_task_description('');
    } 'ValidationError', 'Leere Task-Beschreibung löst ValidationError aus';

    throws_ok {
        validators::validate_task_description('   ');
    } 'ValidationError', 'Nur-Leerzeichen Task löst ValidationError aus';

    my $max_length_task = 'a' x 500;
    lives_ok {
        validators::validate_task_description($max_length_task);
    } 'Task exakt an der Maximallänge (500 Zeichen) wird akzeptiert';

    my $too_long_task = 'a' x 501;
    throws_ok {
        validators::validate_task_description($too_long_task);
    } 'ValidationError', 'Task über 500 Zeichen löst ValidationError aus';
};

# ===========================================================================
# Tests: Modell-Allowlist
# ===========================================================================

subtest 'Modell-Allowlist' => sub {
    for my $model (@validators::ALLOWED_AI_MODELS) {
        lives_ok {
            validators::validate_ai_model_name($model);
        } "Modell aus Allowlist akzeptiert: $model";
    }

    my @invalid_models = (
        'gpt-4',                          # Ohne Provider-Prefix
        'unknown-model',                   # Unbekanntes Modell
        'openrouter/evil; rm -rf /',       # Injection-Versuch
        '',                               # Leer
        'claude-3-5-sonnet-20241022',     # Ohne openrouter/-Prefix
    );

    for my $model (@invalid_models) {
        throws_ok {
            validators::validate_ai_model_name($model);
        } 'ValidationError', "Unbekanntes Modell blockiert: $model";
    }
};

# ===========================================================================
# Tests: Zielsprachen-Validierung
# ===========================================================================

subtest 'Zielsprachen-Validierung' => sub {
    for my $lang (@validators::ALLOWED_TARGET_LANGUAGES) {
        lives_ok {
            validators::validate_target_language($lang);
        } "Sprache akzeptiert: $lang";
    }

    lives_ok {
        validators::validate_target_language('JavaScript');
    } 'Zielsprachen werden auf Kleinschreibung normalisiert';

    throws_ok {
        validators::validate_target_language('cobol');
    } 'ValidationError', 'Nicht unterstützte Sprachen werden blockiert';

    throws_ok {
        validators::validate_target_language('perl5; rm -rf /');
    } 'ValidationError', 'Injection-Versuche in Sprachen-Parameter werden blockiert';
};

# ===========================================================================
# Tests: Timeout-Validierung
# ===========================================================================

subtest 'Timeout-Validierung' => sub {
    my @valid_timeouts = (1, 60, 1800, 3600, 7200);
    for my $timeout (@valid_timeouts) {
        lives_ok {
            validators::validate_timeout_seconds($timeout);
        } "Gültiger Timeout akzeptiert: $timeout";
    }

    my @invalid_timeouts = (0, -1, 7201, 99999);
    for my $timeout (@invalid_timeouts) {
        throws_ok {
            validators::validate_timeout_seconds($timeout);
        } 'ValidationError', "Ungültiger Timeout abgelehnt: $timeout";
    }

    throws_ok {
        validators::validate_timeout_seconds(1800.5);
    } 'ValidationError', 'Float-Timeout (kein Integer) löst ValidationError aus';
};

# ===========================================================================
# Tests: API-Schlüssel-Validierung
# ===========================================================================

subtest 'API-Schlüssel-Validierung' => sub {
    local $ENV{ANTHROPIC_API_KEY} = 'sk-ant-api03-' . ('x' x 30);
    lives_ok {
        validators::load_and_validate_api_key('ANTHROPIC');
    } 'Gültiger API-Schlüssel aus Umgebungsvariable wird zurückgegeben';

    delete $ENV{ANTHROPIC_API_KEY};
    throws_ok {
        validators::load_and_validate_api_key('ANTHROPIC');
    } 'ApiKeyError', 'Fehlende Umgebungsvariable löst ApiKeyError aus';

    local $ENV{ANTHROPIC_API_KEY} = 'short';
    throws_ok {
        validators::load_and_validate_api_key('ANTHROPIC');
    } 'ApiKeyError', 'Zu kurzer API-Schlüssel (< 20 Zeichen) löst ApiKeyError aus';
};

# ===========================================================================
# Tests: File-Change-Detection
# ===========================================================================

subtest 'File-Change-Detection' => sub {
    my $tmp_path = tempdir(CLEANUP => 1);
    my $test_file = File::Spec->catfile($tmp_path, 'script.pl');
    open my $fh, '>', $test_file or die "Kann $test_file nicht öffnen: $!";
    print $fh "print('hello');\n";
    close $fh;

    my $current_hash = create_abstraction::compute_file_sha256($test_file);
    my %hash_cache = ($test_file => $current_hash);

    ok(!create_abstraction::has_source_file_changed($test_file, \%hash_cache), 'Unveränderte Datei wird nicht als geändert erkannt');

    open $fh, '>', $test_file or die "Kann $test_file nicht öffnen: $!";
    print $fh "print('changed!');\n";
    close $fh;

    ok(create_abstraction::has_source_file_changed($test_file, \%hash_cache), 'Geänderte Datei wird korrekt als geändert erkannt');

    my %empty_cache;
    ok(create_abstraction::has_source_file_changed($test_file, \%empty_cache), 'Datei ohne gespeicherten Hash wird als neu/geändert erkannt');

    my $file_a = File::Spec->catfile($tmp_path, 'a.pl');
    my $file_b = File::Spec->catfile($tmp_path, 'b.pl');
    open $fh, '>', $file_a or die "Kann $file_a nicht öffnen: $!";
    print $fh "content A\n";
    close $fh;
    open $fh, '>', $file_b or die "Kann $file_b nicht öffnen: $!";
    print $fh "content B\n";
    close $fh;

    isnt(create_abstraction::compute_file_sha256($file_a), create_abstraction::compute_file_sha256($file_b), 'Verschiedene Dateien haben unterschiedliche SHA-256 Hashes');

    my $identical_content = "identical content";
    open $fh, '>', $file_a or die "Kann $file_a nicht öffnen: $!";
    print $fh $identical_content;
    close $fh;
    open $fh, '>', $file_b or die "Kann $file_b nicht öffnen: $!";
    print $fh $identical_content;
    close $fh;

    is(create_abstraction::compute_file_sha256($file_a), create_abstraction::compute_file_sha256($file_b), 'Identischer Inhalt produziert identischen Hash');
};

# ===========================================================================
# Tests: Atomisches State-File
# ===========================================================================

subtest 'Atomisches State-File' => sub {
    my $state_file = state_file_path();
    my %test_state = (
        file_hashes => { 'script.pl' => 'abc123' },
        last_run => '2026-05-26T10:00:00',
    );

    create_abstraction::save_abstraction_state_atomically($state_file, \%test_state);
    my $loaded_state = create_abstraction::load_abstraction_state($state_file);

    is_deeply($loaded_state, \%test_state, 'State wird korrekt gespeichert und geladen');

    my $nonexistent_path = File::Spec->catfile(tempdir(CLEANUP => 1), 'nonexistent', 'state.json');
    my $result = create_abstraction::load_abstraction_state($nonexistent_path);
    is_deeply($result, {}, 'Fehlende State-Datei gibt leeres Dictionary zurück');

    my $deep_path = File::Spec->catfile(tempdir(CLEANUP => 1), 'a', 'b', 'c', 'state.json');
    create_abstraction::save_abstraction_state_atomically($deep_path, { key => 'value' });
    ok(-f $deep_path, 'Atomisches Schreiben erstellt fehlende Verzeichnisse');

    my $state_path = File::Spec->catfile(tempdir(CLEANUP => 1), 'state.json');
    create_abstraction::save_abstraction_state_atomically($state_path, { data => 1 });
    my @temp_files = glob("$state_path.tmp");
    is(scalar @temp_files, 0, 'Nach erfolgreichem Schreiben gibt es keine .tmp-Datei');

    my $corrupted_state = File::Spec->catfile(tempdir(CLEANUP => 1), 'state.json');
    open my $fh, '>', $corrupted_state or die "Kann $corrupted_state nicht öffnen: $!";
    print $fh "{ invalid json !!!";
    close $fh;

    throws_ok {
        create_abstraction::load_abstraction_state($corrupted_state);
    } 'StateFileError', 'Korrupte State-Datei löst StateFileError aus';
};

# ===========================================================================
# Tests: Sprachzuordnung
# ===========================================================================

subtest 'Sprachzuordnung' => sub {
    for my $language (@validators::ALLOWED_TARGET_LANGUAGES) {
        ok(exists $create_abstraction::LANGUAGE_FILE_EXTENSIONS{$language}, "Jede unterstützte Zielsprache hat eine zugeordnete Datei-Extension: $language");
    }

    my %expected_extensions = (
        perl5 => '.pl',
        perl6 => '.raku',
        javascript => '.js',
        python => '.py',
        bash => '.sh',
        powershell => '.ps1',
        go => '.go',
    );

    for my $language (keys %expected_extensions) {
        is($create_abstraction::LANGUAGE_FILE_EXTENSIONS{$language}, $expected_extensions{$language}, "Korrekte Datei-Extension für $language");
    }
};

done_testing();

File: javascript/TEST_A~1.js
#!/usr/bin/env node
// TEST_A~1.PY — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:abstraction-manager/TEST_A~1.PY
// auch in: Projects@abstractions:abstractions/test_abstractions_manager.py
// Erzeugt: 2026-08-08 durch ABSTRACTIONS_MANAGER.py

/**
 * test_abstractions_manager.js — Unit-Tests für den Abstractions Manager.
 *
 * Testet alle kritischen Sicherheits- und Kernfunktionen:
 *     - Path-Traversal-Schutz (validate_source_file_path)
 *     - Shell-Injection-Prävention (validate_task_description)
 *     - Modell-Allowlist (validate_ai_model_name)
 *     - Zielsprachen-Validierung (validate_target_language)
 *     - Timeout-Validierung (validate_timeout_seconds)
 *     - JSON-Serialisierung (create_abstraction.js Hilfsfunktionen)
 *     - File-Change-Detection (Hash-basiert)
 *     - Atomisches State-File-Schreiben
 *
 * Ausführen:
 *     npm test
 *
 * Author: OpenClaw Team
 * Version: 1.0.0
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const os = require('os');

// Mock für pytest-ähnliche Tests
const assert = require('assert');

// Importiere lokale Module
const {
    ValidationError,
    PortationError,
    ApiKeyError,
    StateFileError,
} = require('./exceptions');

const {
    validate_source_file_path,
    validate_task_description,
    validate_ai_model_name,
    validate_target_language,
    validate_timeout_seconds,
    load_and_validate_api_key,
    ALLOWED_TARGET_LANGUAGES,
    ALLOWED_AI_MODELS,
} = require('./validators');

const {
    compute_file_sha256,
    has_source_file_changed,
    load_abstraction_state,
    save_abstraction_state_atomically,
    LANGUAGE_FILE_EXTENSIONS,
} = require('./create_abstraction');

// ===========================================================================
// Test-Hilfsfunktionen
// ===========================================================================

function createTempDir() {
    return fs.mkdtempSync(path.join(os.tmpdir(), 'test-'));
}

function createTempFile(dir, filename, content) {
    const filepath = path.join(dir, filename);
    fs.writeFileSync(filepath, content, 'utf8');
    return filepath;
}

function removeTempDir(dir) {
    fs.rmSync(dir, { recursive: true, force: true });
}

// ===========================================================================
// Tests: Path-Traversal-Schutz
// ===========================================================================

class TestSourceFilePathValidation {
    static test_valid_path_within_allowed_directory_is_accepted() {
        const tempDir = createTempDir();
        const scriptsDir = path.join(tempDir, 'skills', 'scripts');
        fs.mkdirSync(scriptsDir, { recursive: true });
        const scriptPath = createTempFile(scriptsDir, 'db_maintainer.py', "# Test-Script\nprint('hello world')\n");
        
        process.env.OPENCLAW_WORKSPACE = tempDir;
        
        try {
            const result = validate_source_file_path(scriptPath);
            assert.strictEqual(result, path.resolve(scriptPath));
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_nonexistent_file_raises_file_not_found_error() {
        assert.throws(() => {
            validate_source_file_path('/absolutely/nonexistent/path/script.py');
        }, { code: 'ENOENT' });
    }

    static test_directory_instead_of_file_raises_validation_error() {
        const tempDir = createTempDir();
        try {
            assert.throws(() => {
                validate_source_file_path(tempDir);
            }, ValidationError);
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_path_outside_workspace_raises_validation_error() {
        process.env.OPENCLAW_WORKSPACE = '/home/openclaw/.openclaw/workspace';
        
        const traversalAttempts = [
            '/etc/passwd',
            '/etc/shadow',
            '/root/.ssh/id_rsa',
            '/proc/1/cmdline',
        ];
        
        for (const attempt of traversalAttempts) {
            assert.throws(() => {
                validate_source_file_path(attempt);
            }, (err) => {
                return err instanceof ValidationError || err.code === 'ENOENT';
            });
        }
    }
}

// ===========================================================================
// Tests: Shell-Injection-Prävention
// ===========================================================================

class TestTaskDescriptionValidation {
    static test_valid_task_descriptions_are_accepted() {
        const validTasks = [
            'Port db_maintainer.py to Go',
            'Add error handling to json_processor',
            'Refactor websearch-crawl.sh for Perl 5',
            'Port check-live.js with full JSDoc',
            'Migrate log_collector.py to Ruby',
        ];
        
        for (const task of validTasks) {
            const result = validate_task_description(task);
            assert.strictEqual(result, task.trim());
        }
    }

    static test_shell_metacharacters_are_rejected() {
        const maliciousInputs = [
            ['; rm -rf /', 'Semikolon + Befehl'],
            ['$(cat /etc/passwd)', 'Befehlssubstitution $(...)'],
            ['`whoami`', 'Backtick-Substitution'],
            ['task && evil_cmd', '&&-Verkettung'],
            ['task | cat /etc/shadow', 'Pipe-Redirect'],
            ['task\ncommand', 'Newline-Injection'],
            ['task; exit 0', 'Semikolon-Injection'],
            ['task > /tmp/evil', 'Ausgabe-Redirect'],
        ];
        
        for (const [input, description] of maliciousInputs) {
            assert.throws(() => {
                validate_task_description(input);
            }, ValidationError);
        }
    }

    static test_empty_task_raises_validation_error() {
        assert.throws(() => {
            validate_task_description('');
        }, ValidationError);
    }

    static test_whitespace_only_task_raises_validation_error() {
        assert.throws(() => {
            validate_task_description('   ');
        }, ValidationError);
    }

    static test_task_at_maximum_length_is_accepted() {
        const maxLengthTask = 'a'.repeat(500);
        const result = validate_task_description(maxLengthTask);
        assert.strictEqual(result.length, 500);
    }

    static test_task_exceeding_maximum_length_raises_validation_error() {
        const tooLongTask = 'a'.repeat(501);
        assert.throws(() => {
            validate_task_description(tooLongTask);
        }, ValidationError);
    }
}

// ===========================================================================
// Tests: Modell-Allowlist
// ===========================================================================

class TestAiModelValidation {
    static test_all_allowed_models_are_accepted() {
        for (const model of ALLOWED_AI_MODELS) {
            const result = validate_ai_model_name(model);
            assert.strictEqual(result, model);
        }
    }

    static test_unknown_models_are_rejected() {
        const invalidModels = [
            'gpt-4',                          // Ohne Provider-Prefix
            'unknown-model',                  // Unbekanntes Modell
            'openrouter/evil; rm -rf /',      // Injection-Versuch
            '',                               // Leer
            'claude-3-5-sonnet-20241022',     // Ohne openrouter/-Prefix
        ];
        
        for (const model of invalidModels) {
            assert.throws(() => {
                validate_ai_model_name(model);
            }, ValidationError);
        }
    }
}

// ===========================================================================
// Tests: Zielsprachen-Validierung
// ===========================================================================

class TestTargetLanguageValidation {
    static test_all_supported_languages_are_accepted() {
        for (const language of ALLOWED_TARGET_LANGUAGES) {
            const result = validate_target_language(language);
            assert.strictEqual(result, language.toLowerCase());
        }
    }

    static test_language_is_normalized_to_lowercase() {
        const result = validate_target_language('JavaScript');
        assert.strictEqual(result, 'javascript');
    }

    static test_unsupported_language_raises_validation_error() {
        assert.throws(() => {
            validate_target_language('cobol');
        }, ValidationError);
    }

    static test_language_with_injection_attempt_raises_validation_error() {
        assert.throws(() => {
            validate_target_language('perl5; rm -rf /');
        }, ValidationError);
    }
}

// ===========================================================================
// Tests: Timeout-Validierung
// ===========================================================================

class TestTimeoutValidation {
    static test_valid_timeouts_are_accepted() {
        const validTimeouts = [1, 60, 1800, 3600, 7200];
        for (const timeout of validTimeouts) {
            const result = validate_timeout_seconds(timeout);
            assert.strictEqual(result, timeout);
        }
    }

    static test_out_of_range_timeouts_raise_validation_error() {
        const invalidTimeouts = [0, -1, 7201, 99999];
        for (const timeout of invalidTimeouts) {
            assert.throws(() => {
                validate_timeout_seconds(timeout);
            }, ValidationError);
        }
    }

    static test_float_timeout_raises_validation_error() {
        assert.throws(() => {
            validate_timeout_seconds(1800.5);
        }, ValidationError);
    }
}

// ===========================================================================
// Tests: API-Schlüssel-Validierung
// ===========================================================================

class TestApiKeyValidation {
    static test_valid_api_key_is_returned() {
        const fakeKey = 'sk-ant-api03-' + 'x'.repeat(30);
        process.env.ANTHROPIC_API_KEY = fakeKey;
        const result = load_and_validate_api_key('ANTHROPIC');
        assert.strictEqual(result, fakeKey);
    }

    static test_missing_api_key_raises_api_key_error() {
        delete process.env.ANTHROPIC_API_KEY;
        assert.throws(() => {
            load_and_validate_api_key('ANTHROPIC');
        }, ApiKeyError);
    }

    static test_too_short_api_key_raises_api_key_error() {
        process.env.ANTHROPIC_API_KEY = 'short';
        assert.throws(() => {
            load_and_validate_api_key('ANTHROPIC');
        }, ApiKeyError);
    }
}

// ===========================================================================
// Tests: File-Change-Detection
// ===========================================================================

class TestFileChangeDetection {
    static test_unchanged_file_is_not_detected_as_changed() {
        const tempDir = createTempDir();
        const testFile = path.join(tempDir, 'script.py');
        fs.writeFileSync(testFile, "print('hello')", 'utf8');
        
        try {
            const currentHash = compute_file_sha256(testFile);
            const hashCache = { [testFile]: currentHash };
            
            assert.strictEqual(has_source_file_changed(testFile, hashCache), false);
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_modified_file_is_detected_as_changed() {
        const tempDir = createTempDir();
        const testFile = path.join(tempDir, 'script.py');
        fs.writeFileSync(testFile, "print('hello')", 'utf8');
        const oldHash = compute_file_sha256(testFile);
        
        try {
            const hashCache = { [testFile]: oldHash };
            fs.writeFileSync(testFile, "print('changed!')", 'utf8');
            
            assert.strictEqual(has_source_file_changed(testFile, hashCache), true);
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_new_file_without_cached_hash_is_detected_as_changed() {
        const tempDir = createTempDir();
        const testFile = path.join(tempDir, 'new_script.py');
        fs.writeFileSync(testFile, "print('new')", 'utf8');
        const emptyCache = {};
        
        try {
            assert.strictEqual(has_source_file_changed(testFile, emptyCache), true);
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_different_files_produce_different_hashes() {
        const tempDir = createTempDir();
        const fileA = path.join(tempDir, 'a.py');
        const fileB = path.join(tempDir, 'b.py');
        fs.writeFileSync(fileA, 'content A', 'utf8');
        fs.writeFileSync(fileB, 'content B', 'utf8');
        
        try {
            assert.notStrictEqual(compute_file_sha256(fileA), compute_file_sha256(fileB));
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_identical_content_produces_identical_hash() {
        const tempDir = createTempDir();
        const fileA = path.join(tempDir, 'a.py');
        const fileB = path.join(tempDir, 'b.py');
        const identicalContent = 'identical content';
        fs.writeFileSync(fileA, identicalContent, 'utf8');
        fs.writeFileSync(fileB, identicalContent, 'utf8');
        
        try {
            assert.strictEqual(compute_file_sha256(fileA), compute_file_sha256(fileB));
        } finally {
            removeTempDir(tempDir);
        }
    }
}

// ===========================================================================
// Tests: Atomisches State-File
// ===========================================================================

class TestAtomicStateFile {
    static test_state_is_saved_and_loaded_correctly() {
        const tempDir = createTempDir();
        const stateFile = path.join(tempDir, 'state.json');
        const testState = {
            file_hashes: { 'script.py': 'abc123' },
            last_run: '2026-05-26T10:00:00',
        };
        
        try {
            save_abstraction_state_atomically(stateFile, testState);
            const loadedState = load_abstraction_state(stateFile);
            assert.deepStrictEqual(loadedState, testState);
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_nonexistent_state_file_returns_empty_dict() {
        const tempDir = createTempDir();
        const nonexistentPath = path.join(tempDir, 'nonexistent', 'state.json');
        
        try {
            const result = load_abstraction_state(nonexistentPath);
            assert.deepStrictEqual(result, {});
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_atomic_write_creates_parent_directories() {
        const tempDir = createTempDir();
        const deepPath = path.join(tempDir, 'a', 'b', 'c', 'state.json');
        
        try {
            save_abstraction_state_atomically(deepPath, { key: 'value' });
            assert.strictEqual(fs.existsSync(deepPath), true);
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_atomic_write_leaves_no_temp_file_on_success() {
        const tempDir = createTempDir();
        const statePath = path.join(tempDir, 'state.json');
        
        try {
            save_abstraction_state_atomically(statePath, { data: 1 });
            const tempFiles = fs.readdirSync(tempDir).filter(f => f.endsWith('.tmp'));
            assert.strictEqual(tempFiles.length, 0);
        } finally {
            removeTempDir(tempDir);
        }
    }

    static test_invalid_json_in_state_file_raises_state_file_error() {
        const tempDir = createTempDir();
        const corruptedState = path.join(tempDir, 'state.json');
        fs.writeFileSync(corruptedState, '{ invalid json !!!', 'utf8');
        
        try {
            assert.throws(() => {
                load_abstraction_state(corruptedState);
            }, StateFileError);
        } finally {
            removeTempDir(tempDir);
        }
    }
}

// ===========================================================================
// Tests: Sprachzuordnung
// ===========================================================================

class TestLanguageFileExtensions {
    static test_all_supported_languages_have_file_extensions() {
        for (const language of ALLOWED_TARGET_LANGUAGES) {
            assert.ok(LANGUAGE_FILE_EXTENSIONS.hasOwnProperty(language), 
                `Fehlende Extension für Sprache: ${language}`);
        }
    }

    static test_correct_file_extension_per_language() {
        const languageExtensions = [
            ['perl5', '.pl'],
            ['perl6', '.raku'],
            ['javascript', '.js'],
            ['python', '.py'],
            ['bash', '.sh'],
            ['powershell', '.ps1'],
            ['go', '.go'],
        ];
        
        for (const [language, expectedExt] of languageExtensions) {
            assert.strictEqual(LANGUAGE_FILE_EXTENSIONS[language], expectedExt);
        }
    }
}

// ===========================================================================
// Testausführung
// ===========================================================================

function runTests() {
    const testClasses = [
        TestSourceFilePathValidation,
        TestTaskDescriptionValidation,
        TestAiModelValidation,
        TestTargetLanguageValidation,
        TestTimeoutValidation,
        TestApiKeyValidation,
        TestFileChangeDetection,
        TestAtomicStateFile,
        TestLanguageFileExtensions,
    ];

    let passed = 0;
    let failed = 0;

    for (const testClass of testClasses) {
        const className = testClass.name;
        console.log(`\nRunning tests for ${className}...`);
        
        const methods = Object.getOwnPropertyNames(testClass)
            .filter(prop => prop.startsWith('test_'));
            
        for (const method of methods) {
            try {
                testClass[method]();
                console.log(`  ✓ ${method}`);
                passed++;
            } catch (error) {
                console.log(`  ✗ ${method}: ${error.message}`);
                failed++;
            }
        }
    }

    console.log(`\nTests completed: ${passed} passed, ${failed} failed`);
    process.exit(failed > 0 ? 1 : 0);
}

// Tests ausführen, wenn das Skript direkt aufgerufen wird
if (require.main === module) {
    runTests();
}

module.exports = {
    TestSourceFilePathValidation,
    TestTaskDescriptionValidation,
    TestAiModelValidation,
    TestTargetLanguageValidation,
    TestTimeoutValidation,
    TestApiKeyValidation,
    TestFileChangeDetection,
    TestAtomicStateFile,
    TestLanguageFileExtensions,
};
